### PR TITLE
Add move/remove/swap methods on List

### LINKED ISSUE #3324: Implement Realm.List.{move, swap, remove}
Realm Object Store provides three methods currently not exposed in RJS: move, swap and remove.

References:

* [https://github.com/realm/realm-object-store/blob/v10/src/list.hpp#L83](https://github.com/realm/realm-object-store/blob/v10/src/list.hpp#L83)

* https://github.com/realm/realm-object-store/blob/v10/src/list.hpp#L86

### PR DESCRIPTION
**Pull request by papafe**
This closes #3324

## ☑️ ToDos
* [x] 📝 Changelog entry
* [x] 🚦 Tests

### CODE REVIEW

**Comment by papafe** [comment]
There are a couple of things that I am not sure about, so this is still a draft to get early feedback. 

What kind of behaviour do we want for out of bounds indexes input? It seems that to align on the expectations of JS developers we could have different behaviour for the methods:
- `remove`: it will be a no-op to remove something out of bounds, so that it acts similar to `shift` and `pop`
- `move`/`swap`: those two will throw an exception for out of bounds inputs
Do we agree on this?

I have created a stub for the `list` test suite, looking at what was done for sets and dictionaries. I am unsure of what level of testing do we want to have with this. For instance, do we want to test these methods with multiple kind of list types (int, string, ...) or we just keep it simple?

File: packages/realm/src/List.ts [hunk 1]
```diff
@@ -279,4 +279,38 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
     }
     return result;
   }
+
+  remove(index: number) {
+    assert.inTransaction(this.realm);
+    assert.number(index, "index");
+    const { internal } = this;
```
Comment (kraenhansen) [inline packages/realm/src/List.ts:286]: I'd probably just inline `this.internal` as it's only used twice below.

File: packages/realm/src/List.ts [hunk 2]
```diff
@@ -279,4 +279,38 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
...
+
+  move(from: number, to: number) {
+    assert.inTransaction(this.realm);
+    assert.number(from, "index");
+    assert.number(to, "to");
+
+    //TODO Need to decide if out-of-bounds indexes will be a no-op or an exception
+    const { internal } = this;
+
+    internal.move(from, to);
```
Comment (kraenhansen) [inline packages/realm/src/List.ts:303]: This I would definitely not de-structure, because it makes `internal` appear twice instead of once.

```suggestion
    this.internal.move(from, to);
```

**Review by kraenhansen — COMMENTED** [review]
Great approach, I like it.

Regarding the testing situation, we've talked about codifying a testing strategy and some sort of taxonomy such that we'll increate the likelihood to cover the entire API surface, while not maintaining the same tests twice. Iterating a collection is one example - should that live in a List test or would it make more sense to group that to increase code reuse? Similarly for object creation, should that be tested in an "objects.ts" or "realm.ts"? It creates objects but the `create` method hangs off the `Realm` instance. It would be great if we had principle decisions made on this that we could simply refer to.

The tests you've made this far seems valuable 👍  There are tests being migrated, which our `bindgen` branch will soon be rebased on-top-of: https://github.com/realm/realm-js/blob/test-migration-master/integration-tests/tests/src/tests/list.ts

File: integration-tests/tests/src/tests/list.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+import { expect } from "chai";
+import { Realm } from "realm";
+
+import { openRealmBefore } from "../hooks";
+
+type Item<T = unknown> = { list: Realm.List<T> };
+
+describe.only("List", () => {
+  describe.only("with unconstrained (mixed) values", () => {
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/list.ts:27]: Although convenient, I've seen green tests on CI where there were actually regressions on multiple occasions because someone merged a PR but had forgotten to remove their `.only` statements.

I'd suggest [using `--grep`](https://github.com/realm/realm-js/tree/bindgen/integration-tests#running-the-tests-during-development) instead, as that won't accidentally be committed.

File: integration-tests/tests/src/tests/list.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      });
+
+      expect([...list]).deep.equals([14, 5, 8, 2, 57]);
+
+      this.realm.write(() => {
+        list.swap(2, 1);
+      });
+
+      expect([...list]).deep.equals([14, 8, 5, 2, 57]);
+    });
```
Comment (takameyer) [inline integration-tests/tests/src/tests/list.ts:101]: We should also test what happens when the index is out of range

File: packages/realm/src/List.ts (hunk 2 above)
Comment (takameyer) [reply packages/realm/src/List.ts:303]: It's also being done elsewhere outside of this PR (see `splice` above for example).

**Comment by takameyer** [comment]
Seems correct so far!  Just need to decide what to do with those out of bound indexes.  I think we should throw.

File: packages/realm/src/List.ts (hunk 2 above)
Comment (kraenhansen) [reply packages/realm/src/List.ts:303]: Yeah. And I'm all for that, because it has 9 occurrences of internal it simplifies the code. In this case it doesn't serve that purpose.

File: integration-tests/tests/src/tests/list.ts (hunk 3 above)
Comment (papafe) [reply integration-tests/tests/src/tests/list.ts:27]: Makes sense! I can imagine it happening. 
If that's the case probably it will be worth adding a check on CI too, so it doesn't happen again 😁

File: packages/realm/src/List.ts [hunk 5]
```diff
@@ -279,4 +279,38 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
     }
     return result;
   }
+
+  remove(index: number) {
+    assert.inTransaction(this.realm);
+    assert.number(index, "index");
+    const { internal } = this;
+
+    //TODO I am not sure exactly if we expect a no-op for out of bounds operations or an exception
```
Comment (kneth) [inline packages/realm/src/List.ts:288]: You could argue that out-of-bound should be a no-op as it is more JSish. But as the methods here are not inspired by JS Array, I think we should throw if out of bound.

File: packages/realm/src/List.ts [hunk 6]
```diff
@@ -279,4 +279,38 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
...
+
+    internal.remove(index);
+  }
+
+  move(from: number, to: number) {
+    assert.inTransaction(this.realm);
+    assert.number(from, "index");
+    assert.number(to, "to");
+
+    //TODO Need to decide if out-of-bounds indexes will be a no-op or an exception
```
Comment (kneth) [inline packages/realm/src/List.ts:300]: I am in favor of checking

**Review by kneth — APPROVED** [review]

File: integration-tests/tests/src/tests/list.ts (hunk 4 above)
Comment (papafe) [reply integration-tests/tests/src/tests/list.ts:120]: Just added

File: packages/realm/src/List.ts (hunk 6 above)
Comment (papafe) [reply packages/realm/src/List.ts]: Added

File: packages/realm/src/List.ts [hunk 7]
```diff
@@ -279,4 +279,38 @@ export class List<T = unknown> extends OrderedCollection<T> implements Partially
...
   }
+
+  remove(index: number) {
+    assert.inTransaction(this.realm);
+    assert.number(index, "index");
+    const { internal } = this;
+
+    //TODO I am not sure exactly if we expect a no-op for out of bounds operations or an exception
+    assert(index >= 0, "Index cannot be smaller than 0");
+    assert(index < internal.size, "Index cannot be bigger than the size of the list");
```
[the remark this answers is not in the document]
Comment (papafe) [reply packages/realm/src/List.ts]: Fixed

File: packages/realm/src/List.ts (hunk 2 above)
Comment (papafe) [reply packages/realm/src/List.ts]: Makes sense. I've removed the destructuring.

**Comment by papafe** [comment]
I have addressed all the comments (I think) and I have added the changelog. This is ready to merge in my opinion

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,7 +1,10 @@
 ## vNext (TBD)
 
 ### Enhancements
-* None
+- Added 3 new methods on lists ([#3324](https://github.com/realm/realm-js/issues/3324)):
+  - `list.remove(index)`: removes the element of the list at the specified index.
+  - `list.move(from, to)`: moves one element of the list from one index to another.
+  - `list.swap(index1, index2)`: swaps the positions of the elements of the list at two indexes.
```
Comment (takameyer) [inline CHANGELOG.md:7]:
```suggestion
* Added 3 new methods on lists ([#3324](https://github.com/realm/realm-js/issues/3324)):
  * `list.remove(index)`: removes the element of the list at the specified index.
  * `list.move(from, to)`: moves one element of the list from one index to another.
  * `list.swap(index1, index2)`: swaps the positions of the elements of the list at two indexes.
```

**Review by takameyer — APPROVED** [review]
Looks great :)

File: CHANGELOG.md (hunk 8 above)
Comment (takameyer) [reply CHANGELOG.md:7]: Just to be consistant.  I think at some point we landed on "*" for bullets in the changelog.
